### PR TITLE
Correcting wrong amplitudes from sampleImage

### PR DESCRIPTION
**Pull request by lucadimascolo**
I am running `sampleImage` on CPU, and I found that it may be possible that the amplitudes of the generated visibilities do not correctly reproduce the values of the input model. A comparison of the amplitudes from the FFT of the input image, CASA `ft` and `sampleImage` shows that galario is somehow downscaling their values by a factor between 1 and 2 systematically.
It is worth saying that this becomes significant only in the case of large offsets with respect to the phase center or for extended signals (being the comparison between models, I don't think it is a W-projection-related issue). On the other hand, this does not arise when `sampleImage` is used to create model visibilities with UV coordinates on the same regular grid of the FFT (from `fftfreq`, in the case of numpy), independently of the input model.

This change seems to solve (or at least mitigate) the issue, while producing same results in all the well-behaved cases.

### CODE REVIEW

**Comment by mtazzari** [comment]
Hi Luca,

I don't fully understand the issue here. Can you provide a plot or more quantitative estimate of the discrepancy?
To compute visibilities for sources far from the phase center one needs to apply the primary beam correction. As we wrote in the paper,  galario *does not* apply the primary beam correction (assumes the input Image is primary-beam corrected). 
Therefore, in the cases in which it becomes relevant, the user must apply the PB correction before using `sampleImage()` or `chi2Image()`. E.g. by multiplying the model 2D image by the image of the PB. Note that in this case you cannot use `dRA`, `dDec` options because galario performs the shift by applying a complex phase and therefore you need to provide a model image that is already displaced from the phase center.
See Section 3.2 in the paper https://arxiv.org/abs/1709.06999

**Comment by pavesiriccardo** [comment]
Hey Marco and Luca,
no, I found the same issue independently as well. It's not related to the Primary beam, that's an independent effect and is not related to the interferometry but just to the individual antenna response. No, what I have found is that, you can check this very easily, if you run sampleImage on a model with a source far off the image center, then the flux is not recovered properly (for example if I image the visibilities so produced, the flux in the image is way too low).
Example:  a small Gaussian model, only ~2-3 beams in size, at pixel (275,127) of a 512x512 image, comes out wrong, you can verify for yourself or let me know and I can send you the files.

**Comment by mtazzari** [comment]
Hi Riccardo,
thanks for the details. Can you post/send the code you used to compute the visibilities? I'd be useful to see how `sampleImage()` was used, how `nxy` and `dxy` were chosen given the distribution of uv points. I'd be useful to have also min and max uv distances.

**Comment by pavesiriccardo** [comment]
Ok,
uu=uu.copy(order='C')
vv=vv.copy(order='C')
modelimage=np.roll(np.flip(modelimage,axis=0),1,axis=0).copy(order='C')
model_complex = sampleImage(modelimage, 4.8481368110957469e-07, uu, vv)

where uu and vv where read from a UVFITS file, I've had to change to C-order for some reason, or it would complain. I doubt it's related though...
My minimum UVdist is 20690.7, and max is 412497.0  in wavelengths.
Let me know what else I can help with! Did lucadimascolo fix this already, by any chance? Ive seen some commits, but I haven't tested the latest versions

**Comment by lucadimascolo** [comment]
Here is a plot from a quick test using `py_sampleImage` and the corresponding `py_sampleImageMod` with the modified interpolation scheme (amp-real-imag). As a comparison, I am including the results of the interpolation performed using `int_bilin_MT` (as in `test_galario.py`).

The _fft_ amplitude (blue line) is calculated directly from the `numpy.fft.rfft2`-transformed input image.

As an input I am using a 640x640 image of a small gaussian with a standard deviation of 5 pixels, and a pixel size of 5e-7 radians. The plot shows the amplitude for two cases: 
- Left: gaussian at pixel [320,320], the center of the image
- Right: gaussian at pixel [161,535]

The visibilities are generated considering random points (as in `generate_random_vis`) in the range of UV distances between 20 klambda and 400klambda.

screen shot 2018-05-14 at 13 55 35

Same results can be found also in the case of wider gaussians, a point source (single pixel) or for  extended model.

**Comment by pavesiriccardo** [comment]
so are you saying that you fixed the problem by interpolating amplitude and phases rather than real and imaginary parts?
If that's the correct solution then the code should not calculate the interpolation of real and imaginary parts anymore, those are wasted CPU cycles, am I wrong?

**Comment by lucadimascolo** [comment]
I am actually interpolating real and imaginary parts as well as the amplitudes. Interpolation of phases turned out to be quite unstable (probably because they may vary significantly between neighbour pixels), but they can be computed from the interpolated real and imaginary parts.

**Comment by lucadimascolo** [comment]
It definitely should. I think the strong fluctuations may be due to how phases are defined. One generally computes them in the range [0,2pi) (or [-pi,pi)). For a source at position `(x,y)`, this means that the interpolation, instead of the standard argument `2pi(ux+vy)`, will operate on phase values given by `2pi(ux+vy) % 2pi`, which may introduce large jumps.

**Comment by mtazzari** [comment]
**tl;dr:** In the future, the introduction of a separable kernel convolution algorithm can be evaluated to solve this issue. Currently, I suggest to go ahead with the solution proposed by Luca, which is robust and delivers the most accurate result, but introduces a significant performance loss (see timings below).
@lucadimascolo any idea on how we can make the modified algorithm faster? I tried to see if we can get rid of at least part of the above interpolation of Real and Imaginary part to get the phase but I wasn't successful. 

**Details for the records**
I have finally had time to look into this more carefully. I can reproduce the behaviour Luca explained so far, namely that:

 - Amplitude is smoothly varying across pixels, while Phase is highly fluctuating
 - Real part is highly fluctuating, while Imaginary part is smoothly varying.

Given that, I agree that a  good solution is interpolating Amplitude directly, and Phase indirectly by interpolating Real and Imaginary part and then taking the Phase of the resulting complex number.

This issue is the so called _spectral leakage_ that arises when interpolating the output of DFT on a non-regular grid of frequencies (see e.g. this [SO question](https://dsp.stackexchange.com/questions/29509/extracting-accurate-phase-and-amplitude-information-from-fft-with-an-arbitrary-n)). This is exacerbated when the source is far from the center, but I agree that the discrepancy present also for modest distances from the center is not acceptable.

A possible solution to this issue is interpolating via convolution with a kernel, which is essentially the classical approach for gridding. We wanted to avoid convolution because convolutions are non local operations that are typically very slow. Perhaps on the GPU there is a fast solution with separable kernels (for future reference, see the dedicated [CUDA example](https://docs.nvidia.com/cuda/cuda-samples/index.html#cuda-separable-convolution), but we wanted a symmetric implementation for the CPU at least in first instance so we opted for bilinear interpolation. 

I tested whether the discrepancy is due to a wrong choice of `nxy`, `dxy`. I found that even with proper choice of `nxy`, `dxy` as returned by `get_image_size()` for the given uvpoints, the issue is present. Below I attach a few plots for different source locations and image sizes. The discrepancy increases with increasing distance from the center. 

**Performance**
I tested the modified version in the currently latest commit by Luca 8942f1f and I found that the timing of the `interpolate_core` increases significantly compared to published baseline timings.
I carried out speed benchmark on the CPU only, needs to be tested on the GPU.


| size  	|  t_original (ms)	| t_modified  (ms)| loss |
|---	|---	|---	|---
| 512  	|  40.4398 	        |   177.331 	|   4.4x  |
| 1024  	|  42.873 	        |   180.188	|   4.3x  |
| 2048  	|  45.3321 	        |   180.703 	|   4x     |
| 4096  	|  69.9503 	        |   209.933 	|   3x     |
| 8192  	|  99.0928 	        |   234.806 	|   2.3    |
| 16384  	|  105.811 	        |   243.108 	|   2.3    |






For 512x512 matrix:
testlocation_gauss_25_51 plots
testlocation_gauss_25_51 points
testlocation_gauss_128_256 plots
testlocation_gauss_128_256 points
testlocation_gauss_256_512 plots
testlocation_gauss_256_512 points

For 1024x1024:
testlocation_gauss_51_102 plots
testlocation_gauss_51_102 points

With setup code:
```py
naxis   = 1024
cdelt   = (5e-7)/2.

u = np.linspace(2e4, 4.5e5, 20)
v = u.copy()

f_min = 6
nxy, dxy = get_image_size(u, v, f_min=f_min, f_max=5, verbose=True)
uvdist = np.hypot(u, v)
MRS = 0.6 / np.min(uvdist)
duv = 1 / MRS / f_min

print(check_image_size(u, v, naxis, cdelt, duv, verbose=True))  # test passes
```

File: src/galario.cpp [hunk 1]
```diff
@@ -677,12 +677,29 @@ inline dcomplex interpolate_core(int const nrow, int const ncol, const dcomplex
     dcomplex const final_add2 = CMPLXADD(term2, term3);
     dcomplex const final_add1 = CMPLXADD(term1, final_add2);
 
-    dcomplex interpolated = CMPLXADD(final_add1, y0);
+    dcomplex phaseInterp = CMPLXADD(final_add1, y0);
```
Comment (mtazzari) [inline src/galario.cpp]: Please change the name of this variable, it's confusing since it's a complex number, not only the phase.
I would call it, e.g. `interpolated_ReIm` so that `angleInterp` below can be properly renamed `phaseInterp`.

File: src/galario.cpp [hunk 2]
```diff
@@ -677,12 +677,29 @@ inline dcomplex interpolate_core(int const nrow, int const ncol, const dcomplex
...
 
     if (u < 0.) {
-        interpolated = CMPLXCONJ(interpolated);
+        phaseInterp = CMPLXCONJ(phaseInterp);
     }
 
+    dreal angleInterp = arg(phaseInterp);
```
Comment (mtazzari) [inline src/galario.cpp]: Please rename `angleInterp` to `phaseInterp`

File: src/galario.cpp [hunk 3]
```diff
@@ -677,12 +677,29 @@ inline dcomplex interpolate_core(int const nrow, int const ncol, const dcomplex
...
+    dreal const y1r = abs(y1);
+    dreal const y2r = abs(y2);
+    dreal const y3r = abs(y3);
+
+    dreal ampInterp = y0r;
+    ampInterp += (y3r-y0r)*qr;
+    ampInterp += (y1r-y0r)*tr;
+    ampInterp += (y0r-y1r+y2r-y3r)*tr*qr;
+
+    dcomplex interpolated = dcomplex{ampInterp*dreal(cos(angleInterp)), ampInterp*dreal(sin(angleInterp))};
```
Comment (mtazzari) [inline src/galario.cpp]: `angleInterp` -> `phaseInterp`

File: src/galario.cpp [hunk 4]
```diff
@@ -677,12 +677,29 @@ inline dcomplex interpolate_core(int const nrow, int const ncol, const dcomplex
...
-        interpolated = CMPLXCONJ(interpolated);
+        phaseInterp = CMPLXCONJ(phaseInterp);
     }
 
+    dreal angleInterp = arg(phaseInterp);
+
+    dreal const tr = indv - fl_v;
+    dreal const qr = indu - fl_u;
+
+    dreal const y0r = abs(y0);
```
Comment (mtazzari) [inline src/galario.cpp]: The `abs` and `arg` functions do not exist in the CUDA math library and raise errors when compiled for GPU.
Following the pattern used for `CMPLXMUL`, `CMPLXADD`, etc. you should create macros such as `CMPLXABS` and `CMPLXARG` that define these operations for the CUDA and CPU implementation.
For CUDA there are cuCabsf and cuCabs (single and double precision) and for CPU you can use the simple abs.

If you need help with this I can do it, if you enable me to write on this PR (there should be an option on the right of the PR page (see instructions [here](https://help.github.com/articles/committing-changes-to-a-pull-request-branch-created-from-a-fork/) on how to enable me).

Comment (lucadimascolo) [reply src/galario.cpp]: > any idea on how we can make the modified algorithm faster?

I don't know, and I'm actually quite surprised the change in the algorithm is producing such a dramatic performance loss. Do you know if there is any specific part of `interpolate_core` (other than the double interpolation) that is bottlenecking the execution?


> A few changes requested to make the code run on the GPU.

I am adapting the code now, and will commit the changes as soon as possible.

**Comment by lucadimascolo** [comment]
> any idea on how we can make the modified algorithm faster?
I don't know, and I'm actually quite surprised the change in the algorithm is producing such a dramatic performance loss. Do you know if there is any specific part of interpolate_core (other than the double interpolation) that is bottlenecking the execution?

> A few changes requested to make the code run on the GPU.
I am adapting the code now, and will commit the changes as soon as possible.

**Comment by pavesiriccardo** [comment]
Hey guys, the gridding part is very important. Could we try using “natgrid”? It’s supposed to be a lot better than the naive techniques!
http://scipy-cookbook.readthedocs.io/items/Matplotlib_Gridding_irregularly_spaced_data.html <http://scipy-cookbook.readthedocs.io/items/Matplotlib_Gridding_irregularly_spaced_data.html>

http://www.ncarg.ucar.edu//ngmath/natgrid/nnhome.html <http://www.ncarg.ucar.edu//ngmath/natgrid/nnhome.html>

with C++ code here:

https://github.com/likev/natgrid.cpp <https://github.com/likev/natgrid.cpp>

Thoughts?

**Comment by pavesiriccardo** [comment]
I guess, actually, maybe we should use this method?

F.R. Schwab "Optimal gridding of visibility data in radio interferometry”

That would be the way to know the result is correct… But it might be somewhat slower...

**Comment by lucadimascolo** [comment]
@pavesiriccardo Schwab's approach to gridding is great (well, optimal...) for obtaining estimates of gridded Fourier data from a non-regular sample, but we should go the other way round. `interpolate_core` is basically de-gridding the Fourier-space model by interpolating it at the input non-regular *uv* points.

**Comment by pavesiriccardo** [comment]
Yes, so you mean it should be a simpler problem that does not need the full machinery? Alright, there should be something better at capturing fast oscillations though. Like a weighting with an appropriate, and small kernel?

**Comment by mtazzari** [comment]
Yes, as I said some algorithms that solve the spectral leakage implement convolutions with some kernel functions, which is essentially similar to the Schwab's approach, which provides an optimal kernel for this operation.
I think the fix suggested by Luca provides a robust solution and should be merged asap, since the implementation of gridding by convolution will take time. 
I opened issue #148 to remember to do this.

**Comment by pavesiriccardo** [comment]
Thanks, one more request for clarification. I’m not sure I see how bilinear interpolation is different from another set of weights in averaging over the neighboring points. You seem to make a distinction with kernel methods, but a convolution over a 3x3 kernel or something like that is still just a weighted averaging over a few points. Am I wrong on understanding what you’re saying?
Thanks!

**Comment by mtazzari** [comment]
Tested on GPU: compilation and all tests pass!
Need to run a few more performance tests before merging.
Running them between now and Monday.

**Comment by mtazzari** [comment]
OK, I have been able to nail down the origin of the bottleneck.
For a matrix of size `nx=512`, and 1 million sampling points, `interpolate_core` takes usually 1.0ms before this PR.

With this PR, it takes usually 2.15 ms, or even more.

The line 
```cpp
dcomplex interpolated = dcomplex{interp_amp*dreal(cos(interp_phase)), interp_amp*dreal(sin(interp_phase))};
```
introduces 0.85ms, while all the remaining only 0.3ms.

```cpp
dreal interp_phase = CMPLXARG(interp_reim);

    dreal const tr = indv - fl_v;
    dreal const qr = indu - fl_u;

    dreal const y0r = CMPLXABS(y0);
    dreal const y1r = CMPLXABS(y1);
    dreal const y2r = CMPLXABS(y2);
    dreal const y3r = CMPLXABS(y3);

    dreal interp_amp = y0r;
    interp_amp += (y3r-y0r)*qr;
    interp_amp += (y1r-y0r)*tr;
    interp_amp += (y0r-y1r+y2r-y3r)*tr*qr;
```
Need to figure out if we can get that line faster. Just wanted to write this down as I discovered this and now let's think if there is an easy solution.

I have run more extensive performance tests with many configurations from 512 to 16384, with several openmp threads numbers and GPU threads per block. Essentially, for a call to `sampleProfile`:

 - the GPU version has **no** sizeable slow down, which is excellent (on average < 1-2% in total for one call). 

 - the CPU version is slowed down most significantly for `nx=512` and `nx=1024` by roughly 20% in total for one call, by 8-12% in total for `nx=2048`, <4% for `nx=4096` and < 1% for larger `nx`.

It makes sense that the slowdown has larger impact for smaller matrix sizes, which have an overall smaller computing time. Overall these numbers are acceptable to have a much more robust result. I am just wondering if we can find an easy alternative faster way of writing the complex number  `interpolated`. @lucadimascolo any idea?

**Comment by mtazzari** [comment]
The slow part is really the `cos()` and `sin()` computation. 
For future reference, CUDA has a nice `sincos()` function that computes both of them and saves time compared to calling `sin()` and `cos()` separately. See discussion [here](https://devtalk.nvidia.com/default/topic/415904/improving-the-accuracy-of-the-__sincosf-function/
).
However, since there is no counterpart for C++ we will not implement this.

**Comment by lucadimascolo** [comment]
As C++ alternative to `sincos()`, we may think to use `polar()` (see [here](http://www.cplusplus.com/reference/complex/polar/)).
I can give it a try tomorrow morning.

**Comment by pavesiriccardo** [comment]
Aren’t the GSL trig functions faster?

**Comment by mtazzari** [comment]
Hi both, good spot!
I tried `polar()` but makes no difference unfortunately.
Re the GSL: `sin()` and `cos()` are taken from the math module of the C/C++ compiler that is used when compiling galario. In conda, gcc/g++ is used, so the GSL are automatically used. This is true even if one compiles them manually with gcc/g++.

**Comment by mtazzari** [comment]
@lucadimascolo Whenever you are happy with the PR, I am ready to merge.

**Comment by lucadimascolo** [comment]
@mtazzari One last idea. What about the complex `exp()`? Since `polar()` is not working in a smarter, faster way than the single `sin()`/`cos()` case, I assume it won't be that better. It may be interesting to give it a shot, though. 
If also `exp()` fails, I think we are fine. Let's merge.

**Comment by mtazzari** [comment]
Thank you for the suggestion. The complex `exp` seem to compute `sin` and `cos` pretty in the same way we are doing. Moreover, there is no CUDA counterpart and would it require another macro.
I will proceed with the merge.

**Comment by mtazzari** [comment]
Thank you very much Luca and Riccardo for all the time, the discussion and the great work done to improve galario!!! Looking forward to more contributions, suggestions, comments!
I will now create a contributors list and add your names, please let me know if you are not fine with this. It will be in the homepage of the docs.

I will also bump the version to 1.2 and make a new conda release.
